folding-schemes-solidity/src/verifiers/mod.rs

```diff
@@ -93,14 +153,76 @@ mod tests {
             };
             Groth16::<Bn254>::setup(c, &mut rng).unwrap()
         };
-        // we dont specify any pragma values for both verifiers, the pragma from the decider takes over
-        let groth16_template = Groth16Verifier::from(vk, None);
+
+        let g16_data = Groth16Data::from(vk);
+
         let (pk, vk): (ProverKey<G1>, VerifierKey<Bn254>) = KZGSetup::<Bn254>::setup(&mut rng, 5);
-        let kzg10_template = KZG10Verifier::from(&vk, &pk.powers_of_g[..5], None, None);
-        let decider_template = super::templates::Groth16KZG10DeciderVerifier {
-            groth16_verifier: groth16_template,
-            kzg10_verifier: kzg10_template,
+
+        let kzg_data = KzgData::from((vk, Some(pk.powers_of_g[0..3].to_vec())));
+        let mut bytes = vec![];
+        let nova_cyclefold_data = NovaCyclefoldData::from((g16_data, kzg_data));
+
+        nova_cyclefold_data
+            .serialize_protocol_data(&mut bytes)
+            .unwrap();
+        let obtained_nova_cyclefold_data =
+            NovaCyclefoldData::deserialize_protocol_data(bytes.as_slice()).unwrap();
+
+        assert_eq!(nova_cyclefold_data, obtained_nova_cyclefold_data)
+    }
+
+    #[test]
+    fn nova_cyclefold_decider_template_renders() {
+        let mut rng = ark_std::rand::rngs::StdRng::seed_from_u64(test_rng().next_u64());
+        let (x, y, z) = (21, 21, 42);
+        let (_, vk_g16) = {
+            let c = TestAddCircuit::<Fr> {
+                _f: PhantomData,
+                x,
+                y,
+                z,
+            };
+            Groth16::<Bn254>::setup(c, &mut rng).unwrap()
         };
+
+        let (pk, vk_kzg): (ProverKey<G1>, VerifierKey<Bn254>) =
+            KZGSetup::<Bn254>::setup(&mut rng, 5);
```

Comment:
It seems that these lines are reused multiple times (in nova_cyclefold_data_serde_roundtrip, nova_cyclefold_decider_template_renders, nova_cyclefold_decider_template_compiles, test_groth16_verifier_template_renders, test_groth16_verifier_template_compiles, test_groth16_verifier_accepts_and_rejects_proofs, etc)

How would you see abstracting them in a method like
```rust
    fn setup<'a>(
        n: usize,
    ) -> (
        ProverKey<'a, G1>,
        VerifierKey<Bn254>,
        ark_groth16::ProvingKey<Bn254>,
        ark_groth16::VerifyingKey<Bn254>,
        TestAddCircuit<Fr>,
    ) {
        let mut rng = ark_std::rand::rngs::StdRng::seed_from_u64(test_rng().next_u64());
        let (x, y, z) = (21, 21, 42);
        let circuit = TestAddCircuit::<Fr> {
            _f: PhantomData,
            x,
            y,
            z,
        };
        let (g16_pk, g16_vk) = Groth16::<Bn254>::setup(circuit.clone(), &mut rng).unwrap();

        let (kzg_pk, kzg_vk): (ProverKey<G1>, VerifierKey<Bn254>) =
            KZGSetup::<Bn254>::setup(&mut rng, n);
        (kzg_pk, kzg_vk, g16_pk, g16_vk, circuit)
    }
```